test/anvil/entrypointsimulations.test.ts

```diff
@@ -0,0 +1,252 @@
+      await expect(simulateValidation(op1, entryPoint.address))
+        .to.revertedWith('AA14 initCode must return sender')
+    })
+
+    it('should report failure on insufficient verificationGas (OOG) for creation', async () => {
+      const initCode = getAccountInitCode(accountOwner1.address, simpleAccountFactory)
+      const sender = await entryPoint.callStatic.getSenderAddress(initCode).catch(e => e.errorArgs.sender)
+      const op0 = await fillAndSign({
+        initCode,
+        sender,
+        verificationGasLimit: 5e5,
+        maxFeePerGas: 0
+      }, accountOwner1, entryPoint)
+      // must succeed with enough verification gas.
+      await simulateValidation(op0, entryPoint.address, { gas: '0xF4240' })
+
+      const op1 = await fillAndSign({
+        initCode,
+        sender,
+        verificationGasLimit: 1e5,
+        maxFeePerGas: 0
+      }, accountOwner1, entryPoint)
+      await expect(simulateValidation(op1, entryPoint.address, { gas: '0xF4240' }))
+        .to.revertedWith('AA13 initCode failed or OOG')
+    })
+
+    it('should succeed for creating an account', async () => {
+      const sender = await getAccountAddress(accountOwner1.address, simpleAccountFactory)
+      const op1 = await fillAndSign({
+        sender,
+        initCode: getAccountInitCode(accountOwner1.address, simpleAccountFactory)
+      }, accountOwner1, entryPoint)
+      await fund(op1.sender)
+
+      await simulateValidation(op1, entryPoint.address)
+    })
+
+    it('should not call initCode from entrypoint', async () => {
+      // a possible attack: call an account's execFromEntryPoint through initCode. This might lead to stolen funds.
+      const { proxy: account } = await createAccount(ethersSigner, await accountOwner.getAddress(), entryPoint.address)
+      const sender = createAddress()
+      const op1 = await fillAndSign({
+        initCode: hexConcat([
+          account.address,
+          account.interface.encodeFunctionData('execute', [sender, 0, '0x'])
+        ]),
+        sender
+      }, accountOwner, entryPoint)
+      const error = await simulateValidation(op1, entryPoint.address).catch(e => e)
+      expect(error.message).to.match(/initCode failed or OOG/, error)
+    })
+
+    // TODO: this test is impossible to do with the "state override" approach
+    it.skip('should not use banned ops during simulateValidation', async () => {
+      const op1 = await fillAndSign({
+        initCode: getAccountInitCode(accountOwner1.address, simpleAccountFactory),
+        sender: await getAccountAddress(accountOwner1.address, simpleAccountFactory)
+      }, accountOwner1, entryPoint)
+      await fund(op1.sender)
+      await simulateValidation(op1, entryPoint.address, { gas: 10e6 }).catch(e => e)
+      const block = await ethers.provider.getBlock('latest')
+      const hash = block.transactions[0]
+      await checkForBannedOps(hash, false)
+    })
+  })
+
+  // describe('#simulateHandleOp', () => {
```

Comment:
the one thing missing from simulateHandleOp (which is why we need this "Simulation" contract) is to return the actualGasUsed and success/revert reason of actual call.
they are dumped in an event deep inside innerHandleOp, and the only simple way to propagate them out to the calling "simulateHandleOp" is using storage - a big no-no for a real production code, but good enough for simulate function.